KDL based Movement in Cartesian Space

BoWangFromMars opened the pull request:
KDL(Kinematics and Dynamics Library) installation process is introduced in this branch. Based on forward kinematics and inverse kinematics provided by KDL, a demo, which makes the robot move in z axis in Cartesian space, is added in this branch. Please see the detailed information in pdf file "KDL installation and demo.pdf" in the directory './lbr_demos_fri_ros2_cpp/doc'.

BoWangFromMars committed:
KDL tutorial

 tutorial of KDL installation and demo of Cartesian space movement based on KDL, including forward kinematics and inverse kinematics based on KDL

BoWangFromMars committed:
Cartesian space movement

This demo, based on forward kinematics and inverse kinematics provided by KDL, makes the robot move in z axis in Cartesian space. About how to run this demo, please refer to 'KDL installation and demo.pdf' in the directory './lbr_demos_fri_ros2_cpp/doc/KDL installation and demo.pdf'.

BoWangFromMars committed:
Update CMakeLists.txt

BoWangFromMars committed:
Update package.xml

mhubii commented:
thank you for the PR @BoWangFromMars. I would create a development branch so you can re-target the PR against this branch. Then I can make some modifications to the PR on the dev branch.

mhubii commented on lbr_demos/lbr_demos_fri_ros2_cpp/src/cartesian_path_planning_node.cpp:
```diff
@@ -0,0 +1,217 @@
+            phase_ = 0.0; 
+            count_= 0.0;
+
+            std::string urdf_file_path = "/home/nearlab-iiwa/lbr-stack/src/lbr_fri_ros2_stack/lbr_description/urdf/iiwa7/my_iiwa7.urdf"; // path of your robot urdf file
```
would it be possible to load this from the `/robot_description` parameter instead

mhubii commented on lbr_demos/lbr_demos_fri_ros2_cpp/src/cartesian_path_planning_node.cpp:
```diff
@@ -0,0 +1,217 @@
+
+        KDL::Chain chain_; // robot kinematics chain exetracted from robot URDF file
+
+        double amplitude_; // rad
```
don't think these variables are used anywhere

mhubii commented on lbr_demos/lbr_demos_fri_ros2_cpp/src/cartesian_path_planning_node.cpp:
```diff
@@ -0,0 +1,217 @@
+class JointPositionPublisher:public rclcpp::Node
+{
+	private:
+		rclcpp::Publisher<lbr_fri_msgs::msg::LBRPositionCommand>::SharedPtr publisher_;
```
make sure formatting is correct and line limit is 100 characters

mhubii commented on lbr_demos/lbr_demos_fri_ros2_cpp/src/cartesian_path_planning_node.cpp:
```diff
@@ -0,0 +1,217 @@
+using namespace std::chrono_literals; // this namespace to enable we can use 500ms to represent 500 milliseconds in our code.
+using std::placeholders::_1;
+
+lbr_fri_msgs::msg::LBRState current_robot_state; // store current robot state, including measured joint positions
```
can these be stored as members

mhubii commented on lbr_demos/lbr_demos_fri_ros2_cpp/src/cartesian_path_planning_node.cpp:
```diff
@@ -0,0 +1,217 @@
+        {
+            current_robot_state = msg;
+            
+            if(count_ < 200) // in first 2 seconds, to make 'initial_position_command' be 'measured_joint_position'
```
might be better to introduce `is_init` bool or turn `current_robot_state` into unique pointer and check for null

mhubii commented on lbr_demos/lbr_demos_fri_ros2_cpp/src/cartesian_path_planning_node.cpp:
```diff
@@ -0,0 +1,217 @@
+            KDL::JntArray result_joint_positions = KDL::JntArray(chain.getNrOfJoints());
+            lbr_fri_msgs::msg::LBRPositionCommand joint_position_command;
+
+            auto start = std::chrono::high_resolution_clock::now();
```
maybe only measure time on debug

mhubii commented on lbr_demos/lbr_demos_fri_ros2_cpp/src/cartesian_path_planning_node.cpp:
```diff
@@ -0,0 +1,217 @@
+                {
+                    KDL::Frame cartesian_position_command = initial_cartesian_position;
+
+                    cartesian_position_command.p = KDL::Vector(initial_cartesian_position.p.x(), initial_cartesian_position.p.y(), initial_cartesian_position.p.z() + amplitude_ * sin(phase_));
```
oh okay I see, this performs oscillation in Cartesian space, could you maybe add a Cartesian pose topic or even better a Twist topic (to indicate velocity commands) instead? Twist commands might need some code changes

mhubii reviewed (commented):
Reviewed your PR. Could you please amend changes. Not all suggestions might make sense, so feel free to keep things that make sense to you

BoWangFromMars committed:
Update CMakeLists.txt

BoWangFromMars committed:
Create cartesian_pose_node.cpp

The function of this node is to publish Cartesian Pose state of the robot and receive Cartesian Pose Command from other ros nodes.

BoWangFromMars committed:
Create iiwa7.urdf

BoWangFromMars committed:
Create iiwa14.urdf

BoWangFromMars committed:
Create med7.urdf

BoWangFromMars committed:
Create med14.urdf

BoWangFromMars committed:
Delete lbr_demos/lbr_demos_fri_ros2_cpp/doc/KDL installation and demo.pdf

BoWangFromMars committed:
Add files via upload

This file introduces how to install KDL(kinematics and dynamics library) in our system and a demo based on it is given to make the robot run in z direction in Cartesian space.

BoWangFromMars committed:
Update cartesian_pose_node.cpp

BoWangFromMars committed:
Update cartesian_pose_node.cpp

BoWangFromMars committed:
Update cartesian_path_planning_node.cpp

This node publishes Cartesian Pose command.

BoWangFromMars committed:
Update cartesian_path_planning_node.cpp

This node, as a demo, publishes Cartesian Pose command to make the robot run in z direction in Cartesian space.

BoWangFromMars committed:
Update cartesian_pose_node.cpp

This node publishes Cartesian Pose state of the robot and receives Cartesian Pose command from other ros nodes, and transform it into joint positions based on KDL to send to the robot.

BoWangFromMars committed:
Update package.xml

mhubii commented:
so ideally the robot descriptions will be loaded from the `/robot_description` topic, and there is no need for the additional `.urdf` files. Hope that makes sense. Alternatively, the robot description could be parsed as a parameter to the node.

### From Topic
Create subscription to `/robot_description`. `robot_state_publisher` publishes the robot description with QoS `transient local` topic, see e.g. https://docs.ros.org/en/rolling/Concepts/Intermediate/About-Quality-of-Service-Settings.html:

- Durability
  - Transient local: the publisher becomes responsible for persisting samples for “late-joining” subscriptions.

### From Parameter
Refer e.g. https://github.com/lbr-stack/lbr_fri_ros2_stack/blob/036b23b2f91af70e1ca490e0c81660b6b17ae114/lbr_demos/lbr_demos_fri_ros2_advanced_cpp/src/admittance_control_node.cpp#L17

mhubii commented on lbr_description/urdf/iiwa14/iiwa14.urdf:
```diff
@@ -0,0 +1,229 @@
+<?xml version="1.0" ?>
```
these files are redundant

mhubii commented on lbr_demos/lbr_demos_fri_ros2_cpp/src/cartesian_pose_node.cpp:
```diff
@@ -0,0 +1,246 @@
+
+      // get the path of urdf file 
+      std::filesystem::path current_working_directory = std::filesystem::current_path();
+      std::string urdf_file_path = current_working_directory.string() 
```
getting the current path might not work in general

BoWangFromMars replied:
Hello, @mhubii,  thank you very much for your suggestion and QoS documentation, I learn a lot from it. I modify my code following your advice today and then upload it again after I test it.

mhubii replied:
awesome, happy to merge once the `/robot_description` is loaded from topic / parameter and `.urdf` files are removed

mhubii commented on lbr_demos/lbr_demos_fri_ros2_cpp/doc/KDL installation.pdf:
one more question regarding this doc. KDL is usually installed through `rosdep` as listed in `package.xml` files, refer e.g. https://github.com/lbr-stack/lbr_fri_ros2_stack/blob/036b23b2f91af70e1ca490e0c81660b6b17ae114/lbr_demos/lbr_demos_fri_ros2_advanced_cpp/package.xml#L18

mhubii commented on lbr_demos/lbr_demos_fri_ros2_cpp/package.xml:
```diff
@@ -12,6 +12,12 @@
+  <depend>urdf</depend>
+  <depend>geometry_msgs</depend>
+  <depend>sensor_msgs</depend>
+  <depend>orocos_kdl</depend>
```
this dependency should be replaced by `orocos_kdl_vendor`

BoWangFromMars replied:
okay, I get it, thank you.

mhubii reviewed (commented):
some more dependency hints

BoWangFromMars committed:
Delete lbr_description/urdf/iiwa14/iiwa14.urdf

BoWangFromMars committed:
Delete lbr_description/urdf/iiwa7/iiwa7.urdf

BoWangFromMars committed:
Delete lbr_description/urdf/med14/med14.urdf

BoWangFromMars committed:
Delete lbr_description/urdf/med7/med7.urdf

BoWangFromMars committed:
Delete lbr_demos/lbr_demos_fri_ros2_cpp/doc/KDL installation.pdf

BoWangFromMars committed:
Update package.xml

BoWangFromMars committed:
Update CMakeLists.txt

BoWangFromMars committed:
Update CMakeLists.txt

BoWangFromMars committed:
Update cartesian_pose_node.cpp

BoWangFromMars committed:
Add files via upload

mhubii commented:
thank you again for the PR @BoWangFromMars. I merged this into the `dev` branch and will update in #170.

If you want to make amend changes yourself, please add a PR to the dev branch again.
